scripts/Deploy-CustomScriptsExtension.ps1

```diff
@@ -0,0 +1,179 @@
+         )
+
+    
+    # append header to identify where the call came from for debugging purposes
+    if ($Context -ne "")
+    {
+        $Message = "$Context - $Message";
+    }
+
+    # if necessary, prepend a blank line
+    if ($ClearLine -eq $true)
+    {
+        $logTime = [System.Environment]::NewLine
+    }
+
+    # prepend log time
+    $logTime += "[$(get-date -format u)]";
+
+    if($NoNewLine -eq $false -and $SkipTimestamp -eq $false)
+    {
+        $logLine = "$logTime :: $Message";
+    }
+    else
+    {
+        $logLine = $Message;
+    }
+
+    switch($LogType)
+    {
+        "Verbose" {  Write-Verbose $logLine; }
+        "Output"  {  Write-Output $logLine ; }
+        "Host"    {  Write-Host $logLine -ForegroundColor $ForegroundColor -NoNewline:$NoNewLine; }
+        "Error"   {  Write-Error $logLine; }
+        "Warning" {  Write-Warning $logLine ; }
+        default   {  Write-Host $logLine -ForegroundColor $ForegroundColor -NoNewline:$NoNewLine; }
+    }
+}
+
+$jobs = @()
+
+# if upgrade is set, we skip deleting the existing extensions
+if ($Upgrade -eq $false)
+{
+
+    # get all VMs
+    Log-Message "Getting all VMs in the '$ResourceGroupName' resource group";
+    $vms = Get-AzureRmVM -ResourceGroupName $ResourceGroupName -Verbose
+
+
+    # iterate each VM and remove any existing custom scription extension
+    foreach($vm in $vms)
+    {
+        foreach($extension in $vm.Extensions)
+        {
+            # TODO: run this within a job to speed up the removal
+            $extensionName = $extension.Id.split("/") | select -Last 1
+
+             $jobs += Start-Job  -Name "Remove-CSX-$($vm.Name)-$($extensionName)" –Scriptblock {
+                                                    param($vmName, $extensionName, $ResourceGroupName, $AadWebClientAppKey, $AadWebClientId, $AadTenantId, $AzureSubscriptionName)
+
+                                                    # Login
+                                                    $clientSecret = ConvertTo-SecureString -String $AadWebClientAppKey -AsPlainText -Force
+                                                    $aadCredential = New-Object System.Management.Automation.PSCredential($AadWebClientId, $clientSecret)
+                                                    Login-AzureRmAccount -ServicePrincipal -TenantId $AadTenantId -SubscriptionName $AzureSubscriptionName -Credential $aadCredential -ErrorAction Stop
+                                                    Set-AzureSubscription -SubscriptionName $AzureSubscriptionName | Out-Null
+
+                                                    Remove-AzureRmVMCustomScriptExtension -ResourceGroupName $ResourceGroupName -VMName $vmName -Name $extensionName -Force
+                                                
+                                                
+            } -ArgumentList $vm.Name, $extensionName, $ResourceGroupName, $AadWebClientAppKey, $AadWebClientId, $AadTenantId, $AzureSubscriptionName;
```

Comment:
Unfortunately, i don't believe that this style is supported with powershell script blocks.